cocos/core/gfx/render-pass.ts

```diff
@@ -78,4 +80,21 @@ export abstract class GFXRenderPass extends GFXObject {
     public abstract initialize (info: IGFXRenderPassInfo): boolean;
 
     public abstract destroy (): void;
+    public hash(): number { return this._hash; }
```

Comment:
computeHash should be invoked in inherited classes as it know when attributes are set.

Reply:
Are we accessing the hash before attributes are set ? If so, maybe a dirty flag is needed for the hash, I don't think we can ensure that the attributes won't be modified after the hash is set.

Another concern about this is the maintainability, in this implementation, upper level need to ensure the hash exists before accessing it, or need to handle the null hash by manually trigger the hash calculation. Either way, it's hard to maintain for upper level developers

Reply:
This class is an abstract class,  the hash value can be accessed after sub class is created. And the attributes are only set when sub class is created, so that's the correct time to calculate the hash value. And yes, it is set once and can not be modified.

Reply:
> And the attributes are only set when sub class is created, so that's the correct time to calculate the hash value.

Sorry, it is not `attributes`. It is `color attachments and depth/stencil attachment`. You can not modify them after it is created. If you want to use different attachments, you should create a new render pass.